File: xjarchive/xjarchive.py
#!/usr/bin/env python3
"""xjarchive -- reference encoder and decoder.

Implements GRAMMAR.md exactly. This is the reference:
where the code and the grammar disagree, the grammar is right and the code is
a bug.

A node is one of:
    (tag, attrs, children)   an element; attrs is a list of (key, value)
    bytes                    text

Everything is bytes. A xjarchive document is a byte string and the decoder must
not assume UTF-8 -- an encoding that only works for text is one that silently
corrupts binary.

Copyright (c) 2026 M. A. Chatterjee, BSD-2-Clause.
"""

from dataclasses import dataclass, field
from typing import List, Tuple, Union

IMPLICIT_TAG = b"#text"
MAX_DEPTH = 256
VARINT_MAX_GROUPS = 10


class XjarchiveError(ValueError):
    """Malformed input. The decoder rejects rather than guesses (grammar §6)."""


@dataclass
class Element:
    tag: bytes
    attrs: List[Tuple[bytes, bytes]] = field(default_factory=list)
    children: List[Union["Element", bytes]] = field(default_factory=list)


Node = Union[Element, bytes]


# ---------------------------------------------------------------------------
# Encoding
# ---------------------------------------------------------------------------

def _escape(raw: bytes, reserved: bytes, escape_leading: bytes = b"") -> bytes:
    """Escape exactly the reserved bytes and no others (grammar §5, §7 rule 4).

    Over-escaping still round-trips, so it passes naive tests while costing
    bytes -- on Wikipedia text, escaping commas and brackets that never needed
    it cost 888,576 bytes and inverted a comparison.
    """
    out = bytearray()
    for i, b in enumerate(raw):
        ch = bytes([b])
        if ch == b"\\" or ch in _split(reserved) or (i == 0 and ch in _split(escape_leading)):
            out += b"\\" + ch
        else:
            out += ch
    return bytes(out)


def _split(bs: bytes) -> set:
    return {bytes([b]) for b in bs}


def _varint(n: int) -> bytes:
    out = bytearray()
    while True:
        b = n & 0x7F
        n >>= 7
        out.append(b | (0x80 if n else 0))
        if not n:
            return bytes(out)


def _read_varint(buf: bytes, i: int):
    n = shift = groups = 0
    while True:
        if i >= len(buf):
            raise XjarchiveError("truncated varint")
        groups += 1
        if groups > VARINT_MAX_GROUPS:
            raise XjarchiveError("varint longer than 10 groups")
        b = buf[i]
        i += 1
        n |= (b & 0x7F) << shift
        shift += 7
        if not b & 0x80:
            return n, i


def _size_inline(b: bytes) -> int:
    return len(b) + sum(1 for x in b if x in b"{}\\")


def _size_nulterm(b: bytes) -> int:
    return 2 + len(b) + sum(1 for x in b if x in b"\x00\\") + 1


def _size_counted(b: bytes) -> int:
    return 2 + len(_varint(len(b))) + len(b)


def choose_form(payload: bytes, leading_bracket: bool = False) -> str:
    """Grammar §5.1: shortest wins; ties break counted, terminated, inline.

    Mandatory rather than advisory -- three legal encodings of one payload
    would otherwise mean three legal files for one document, and
    byte-identical output is the property everything here rests on.
    """
    c = _size_counted(payload)
    t = _size_nulterm(payload)
    i = _size_inline(payload) + (1 if leading_bracket and payload[:1] == b"[" else 0)
    best = min(c, t, i)
    if c == best:
        return "counted"
    if t == best:
        return "terminated"
    return "inline"


def _encode_binary(payload: bytes, form: str) -> bytes:
    if form == "counted":
        return b"\\B" + _varint(len(payload)) + payload
    if form == "terminated":
        body = payload.replace(b"\\", b"\\\\").replace(b"\x00", b"\\0")
        return b"\\b" + body + b"\x00"
    raise ValueError(form)


def _merge_text(children: List[Node]) -> List[Node]:
    """Canonical form §7 rule 5: nothing separates adjacent text on the wire,
    so ["a", "b"] and ["ab"] are the same document. Merge on the way out, and
    the decoder's output is then well defined."""
    out: List[Node] = []
    for c in children:
        if isinstance(c, (bytes, bytearray)) and out and isinstance(out[-1], (bytes, bytearray)):
            out[-1] = bytes(out[-1]) + bytes(c)
        else:
            out.append(c)
    return [c for c in out if not (isinstance(c, (bytes, bytearray)) and not c)]


def encode(node: Node, depth: int = 0) -> bytes:
    """Encode one node into canonical xjarchive (grammar §7)."""
    if depth > MAX_DEPTH:
        raise XjarchiveError(f"nesting deeper than {MAX_DEPTH}")

    if isinstance(node, (bytes, bytearray)):
        # Bare text at document level is a singleton element.
        return b"{" + _escape(bytes(node), b"{}", b"[") + b"}"

    tag = _escape(node.tag, b",{}")
    out = bytearray(b"{" + tag)

    if node.attrs:
        pairs = [
            _escape(k, b":,]") + b":" + _escape(v, b",]")
            for k, v in node.attrs
        ]
        # §7 rule 2: encoders never write the optional trailing comma.
        out += b",[" + b",".join(pairs) + b"]"

    children = _merge_text(node.children)
    if children:
        out += b","
        for i, child in enumerate(children):
            if isinstance(child, (bytes, bytearray)):
                # §4.2: only a leading '[' needs escaping, and only when it is
                # the first byte of the content slot.
                at_start = i == 0 and not node.attrs
                payload = bytes(child)
                form = choose_form(payload, at_start)
                if form == "inline":
                    out += _escape(payload, b"{}", b"[" if at_start else b"")
                else:
                    out += _encode_binary(payload, form)
            else:
                out += encode(child, depth + 1)
    elif node.tag != IMPLICIT_TAG:
        # §4.3: an element with a tag always has a comma, so that {foo} stays
        # available for the singleton.
        out += b","

    return bytes(out + b"}")


def encode_document(nodes: List[Node]) -> bytes:
    return b"".join(encode(n) for n in nodes)


# ---------------------------------------------------------------------------
# Decoding
# ---------------------------------------------------------------------------

class _Reader:
    def __init__(self, buf: bytes):
        self.buf = buf
        self.i = 0

    def eof(self) -> bool:
        return self.i >= len(self.buf)

    def peek(self) -> bytes:
        if self.eof():
            raise XjarchiveError("unexpected end of input")
        return self.buf[self.i : self.i + 1]

    def take(self) -> bytes:
        ch = self.peek()
        self.i += 1
        return ch

    def expect(self, ch: bytes) -> None:
        got = self.take()
        if got != ch:
            raise XjarchiveError(f"expected {ch!r} at offset {self.i - 1}, got {got!r}")


# Every reserved byte, per grammar §2: a byte reserved in one position must be
# escapable there, and the decoder cannot tell which position an escape came
# from without accepting them all.
_VALID_ESCAPES = _split(b"{}[],:\\")


def _read_escaped(r: _Reader, stop: bytes) -> bytes:
    """Read until an unescaped byte in `stop`. Does not consume the stop byte."""
    out = bytearray()
    while True:
        if r.eof():
            raise XjarchiveError("unterminated element")
        ch = r.peek()
        if ch == b"\\":
            r.take()
            nxt = r.take()
            if nxt not in _VALID_ESCAPES:
                # §6: an unknown escape means a version mismatch, not a literal.
                raise XjarchiveError(f"invalid escape \\{nxt.decode('latin-1')}")
            out += nxt
            continue
        if ch in _split(stop):
            return bytes(out)
        out += r.take()


def _read_binary(r: _Reader) -> bytes:
    r.expect(b"\\")
    kind = r.take()
    if kind == b"B":
        n, i = _read_varint(r.buf, r.i)
        # §6: validate the length against what is left BEFORE allocating. A
        # length read from the file and trusted is how a decoder is made to
        # abort on a ten-byte input.
        if n > len(r.buf) - i:
            raise XjarchiveError(f"binary run claims {n} bytes, {len(r.buf) - i} remain")
        r.i = i + n
        return r.buf[i : i + n]

    out = bytearray()
    while True:
        if r.eof():
            raise XjarchiveError("unterminated binary run")
        ch = r.take()
        if ch == b"\x00":
            return bytes(out)
        if ch == b"\\":
            nxt = r.take()
            if nxt == b"0":
                out += b"\x00"
            elif nxt == b"\\":
                out += b"\\"
            else:
                raise XjarchiveError(f"invalid escape in binary run: \\{nxt.decode('latin-1')}")
            continue
        out += ch


def _decode_element(r: _Reader, depth: int) -> Element:
    if depth > MAX_DEPTH:
        raise XjarchiveError(f"nesting deeper than {MAX_DEPTH}")
    r.expect(b"{")

    head = _read_escaped(r, b",{}")
    ch = r.peek()

    if ch == b"}":
        # §4.3: no comma -> singleton, content with the implicit tag.
        r.take()
        return Element(IMPLICIT_TAG, [], [head] if head else [])

    if ch == b"{":
        # A tag immediately followed by a child, with no comma, is not legal.
        raise XjarchiveError("element with children must have a comma after the tag")

    r.expect(b",")
    el = Element(head)

    # §4.2: attributes iff the byte after the comma is '['.
    if not r.eof() and r.peek() == b"[":
        r.take()
        if r.peek() != b"]":
            while True:
                key = _read_escaped(r, b":,]")
                if r.peek() != b":":
                    raise XjarchiveError("attribute pair without ':'")
                r.take()
                val = _read_escaped(r, b",]")
                el.attrs.append((key, val))
                if r.peek() == b",":
                    r.take()
                    if r.peek() == b"]":   # §7 rule 2: accepted, never emitted
                        break
                    continue
                break
        r.expect(b"]")
        if r.peek() == b",":
            r.take()
        elif r.peek() == b"}":
            r.take()
            return el

    # Content: text and child elements, in order.
    while True:
        if r.eof():
            raise XjarchiveError("unterminated element")
        ch = r.peek()
        if ch == b"}":
            r.take()
            return el
        if ch == b"{":
            el.children.append(_decode_element(r, depth + 1))
            continue
        if ch == b"\\" and r.buf[r.i + 1 : r.i + 2] in (b"b", b"B"):
            # §5.1. A letter after the escape byte introduces a command; the
            # reserved bytes are all punctuation, so the two cannot collide.
            el.children.append(_read_binary(r))
            continue
        text = _read_escaped(r, b"{}")
        if text:
            el.children.append(text)


def decode(buf: bytes) -> Element:
    r = _Reader(buf)
    el = _decode_element(r, 0)
    if not r.eof():
        raise XjarchiveError(f"trailing bytes at offset {r.i}")
    return el


def decode_document(buf: bytes) -> List[Element]:
    r = _Reader(buf)
    out = []
    while not r.eof():
        out.append(_decode_element(r, 0))
    return out

File: xjarchive/xjarchive.js
'use strict';
/**
 * xjarchive -- reference encoder and decoder (JavaScript).
 *
 * Implements GRAMMAR.md exactly. Where this code and the grammar disagree,
 * the grammar is right and this is a bug.
 *
 * A node is either an Element, or a Uint8Array of text/binary. Everything is
 * bytes: a xjarchive document is a byte string and the decoder must not assume
 * UTF-8, because an encoding that only works for text is one that silently
 * corrupts binary.
 *
 * This is deliberately a second independent implementation rather than a
 * binding over the Python one. Two implementations that can disagree, checked
 * against shared vectors, is the only way to find out whether the grammar is
 * actually unambiguous.
 *
 * Copyright (c) 2026 M. A. Chatterjee, BSD-2-Clause.
 */

const IMPLICIT_TAG = Buffer.from('#text');
const MAX_DEPTH = 256;
const VARINT_MAX_GROUPS = 10;

class XjarchiveError extends Error {}

class Element {
    constructor(tag, attrs = [], children = []) {
        this.tag = Buffer.from(tag);
        this.attrs = attrs.map(([k, v]) => [Buffer.from(k), Buffer.from(v)]);
        this.children = children.map((c) => (c instanceof Element ? c : Buffer.from(c)));
    }
}

// --- bytes ---------------------------------------------------------------

const BSLASH = 0x5c;
const NUL = 0x00;

function isSet(set, byte) {
    return set.indexOf(byte) !== -1;
}

/** Escape exactly the reserved bytes and no others (grammar §5, §7 rule 4). */
function escapeBytes(raw, reserved, escapeLeading) {
    const out = [];
    for (let i = 0; i < raw.length; i++) {
        const b = raw[i];
        const lead = i === 0 && escapeLeading !== undefined && b === escapeLeading;
        if (b === BSLASH || isSet(reserved, b) || lead) out.push(BSLASH);
        out.push(b);
    }
    return Buffer.from(out);
}

function varint(n) {
    const out = [];
    for (;;) {
        const b = n & 0x7f;
        n = Math.floor(n / 128);
        out.push(n ? b | 0x80 : b);
        if (!n) return Buffer.from(out);
    }
}

function readVarint(buf, i) {
    let n = 0;
    let shift = 1;
    let groups = 0;
    for (;;) {
        if (i >= buf.length) throw new XjarchiveError('truncated varint');
        if (++groups > VARINT_MAX_GROUPS) throw new XjarchiveError('varint longer than 10 groups');
        const b = buf[i++];
        n += (b & 0x7f) * shift;
        shift *= 128;
        if (!(b & 0x80)) return [n, i];
    }
}

// --- which binary form (grammar §5.1) ------------------------------------

function sizeInline(b) {
    let n = b.length;
    for (const x of b) if (x === 0x7b || x === 0x7d || x === BSLASH) n++;
    return n;
}

function sizeTerminated(b) {
    let n = 2 + b.length + 1;
    for (const x of b) if (x === NUL || x === BSLASH) n++;
    return n;
}

function sizeCounted(b) {
    return 2 + varint(b.length).length + b.length;
}

/**
 * Shortest wins; ties break counted, terminated, inline. Mandatory rather
 * than advisory -- three legal encodings of one payload would otherwise mean
 * three legal files for one document.
 */
function chooseForm(payload, leadingBracket = false) {
    const c = sizeCounted(payload);
    const t = sizeTerminated(payload);
    const i =
        sizeInline(payload) + (leadingBracket && payload.length && payload[0] === 0x5b ? 1 : 0);
    const best = Math.min(c, t, i);
    if (c === best) return 'counted';
    if (t === best) return 'terminated';
    return 'inline';
}

function encodeBinary(payload, form) {
    if (form === 'counted') {
        return Buffer.concat([Buffer.from('\\B'), varint(payload.length), payload]);
    }
    const body = [];
    for (const b of payload) {
        if (b === BSLASH) body.push(BSLASH, BSLASH);
        else if (b === NUL) body.push(BSLASH, 0x30); // \0
        else body.push(b);
    }
    return Buffer.concat([Buffer.from('\\b'), Buffer.from(body), Buffer.from([NUL])]);
}

// --- encoding ------------------------------------------------------------

/** §7 rule 5: nothing separates adjacent text on the wire. */
function mergeText(children) {
    const out = [];
    for (const c of children) {
        if (!(c instanceof Element) && out.length && !(out[out.length - 1] instanceof Element)) {
            out[out.length - 1] = Buffer.concat([out[out.length - 1], c]);
        } else {
            out.push(c);
        }
    }
    return out.filter((c) => c instanceof Element || c.length > 0);
}

const R_TAG = [0x2c, 0x7b, 0x7d];
const R_KEY = [0x3a, 0x2c, 0x5d];
const R_VAL = [0x2c, 0x5d];
const R_TEXT = [0x7b, 0x7d];

function encode(node, depth = 0) {
    if (depth > MAX_DEPTH) throw new XjarchiveError(`nesting deeper than ${MAX_DEPTH}`);

    if (!(node instanceof Element)) {
        return Buffer.concat([
            Buffer.from('{'),
            escapeBytes(node, R_TEXT, 0x5b),
            Buffer.from('}'),
        ]);
    }

    const parts = [Buffer.from('{'), escapeBytes(node.tag, R_TAG)];

    if (node.attrs.length) {
        const pairs = node.attrs.map(([k, v]) =>
            Buffer.concat([escapeBytes(k, R_KEY), Buffer.from(':'), escapeBytes(v, R_VAL)])
        );
        parts.push(Buffer.from(',['));
        pairs.forEach((p, i) => {
            if (i) parts.push(Buffer.from(','));
            parts.push(p);
        });
        parts.push(Buffer.from(']'));
    }

    const children = mergeText(node.children);
    if (children.length) {
        parts.push(Buffer.from(','));
        children.forEach((child, i) => {
            if (child instanceof Element) {
                parts.push(encode(child, depth + 1));
            } else {
                const atStart = i === 0 && !node.attrs.length;
                const form = chooseForm(child, atStart);
                parts.push(
                    form === 'inline'
                        ? escapeBytes(child, R_TEXT, atStart ? 0x5b : undefined)
                        : encodeBinary(child, form)
                );
            }
        });
    } else if (!node.tag.equals(IMPLICIT_TAG)) {
        parts.push(Buffer.from(','));
    }

    parts.push(Buffer.from('}'));
    return Buffer.concat(parts);
}

function encodeDocument(nodes) {
    return Buffer.concat(nodes.map((n) => encode(n)));
}

// --- decoding ------------------------------------------------------------

const VALID_ESCAPES = [0x7b, 0x7d, 0x5b, 0x5d, 0x2c, 0x3a, BSLASH];

class Reader {
    constructor(buf) {
        this.buf = buf;
        this.i = 0;
    }
    eof() {
        return this.i >= this.buf.length;
    }
    peek() {
        if (this.eof()) throw new XjarchiveError('unexpected end of input');
        return this.buf[this.i];
    }
    take() {
        const b = this.peek();
        this.i++;
        return b;
    }
    expect(byte) {
        const got = this.take();
        if (got !== byte) throw new XjarchiveError(`expected ${byte} at ${this.i - 1}, got ${got}`);
    }
}

function readEscaped(r, stop) {
    const out = [];
    for (;;) {
        if (r.eof()) throw new XjarchiveError('unterminated element');
        const ch = r.peek();
        if (ch === BSLASH) {
            r.take();
            const nxt = r.take();
            if (!isSet(VALID_ESCAPES, nxt)) {
                throw new XjarchiveError(`invalid escape \\${String.fromCharCode(nxt)}`);
            }
            out.push(nxt);
            continue;
        }
        if (isSet(stop, ch)) return Buffer.from(out);
        out.push(r.take());
    }
}

function readBinary(r) {
    r.expect(BSLASH);
    const kind = r.take();

    if (kind === 0x42) {
        // 'B'
        const [n, i] = readVarint(r.buf, r.i);
        // §6: validate the length against what remains BEFORE allocating.
        if (n > r.buf.length - i) {
            throw new XjarchiveError(`binary run claims ${n} bytes, ${r.buf.length - i} remain`);
        }
        r.i = i + n;
        return r.buf.subarray(i, i + n);
    }

    const out = [];
    for (;;) {
        if (r.eof()) throw new XjarchiveError('unterminated binary run');
        const ch = r.take();
        if (ch === NUL) return Buffer.from(out);
        if (ch === BSLASH) {
            const nxt = r.take();
            if (nxt === 0x30) out.push(NUL);
            else if (nxt === BSLASH) out.push(BSLASH);
            else throw new XjarchiveError(`invalid escape in binary run: \\${String.fromCharCode(nxt)}`);
            continue;
        }
        out.push(ch);
    }
}

function decodeElement(r, depth) {
    if (depth > MAX_DEPTH) throw new XjarchiveError(`nesting deeper than ${MAX_DEPTH}`);
    r.expect(0x7b); // {

    const head = readEscaped(r, [0x2c, 0x7b, 0x7d]);
    const ch = r.peek();

    if (ch === 0x7d) {
        r.take();
        return new Element(IMPLICIT_TAG, [], head.length ? [head] : []);
    }
    if (ch === 0x7b) {
        throw new XjarchiveError('element with children must have a comma after the tag');
    }

    r.expect(0x2c); // ,
    const el = new Element(head);

    if (!r.eof() && r.peek() === 0x5b) {
        r.take();
        if (r.peek() !== 0x5d) {
            for (;;) {
                const key = readEscaped(r, [0x3a, 0x2c, 0x5d]);
                if (r.peek() !== 0x3a) throw new XjarchiveError("attribute pair without ':'");
                r.take();
                const val = readEscaped(r, [0x2c, 0x5d]);
                el.attrs.push([key, val]);
                if (r.peek() === 0x2c) {
                    r.take();
                    if (r.peek() === 0x5d) break;
                    continue;
                }
                break;
            }
        }
        r.expect(0x5d);
        if (r.peek() === 0x2c) r.take();
        else if (r.peek() === 0x7d) {
            r.take();
            return el;
        }
    }

    for (;;) {
        if (r.eof()) throw new XjarchiveError('unterminated element');
        const c = r.peek();
        if (c === 0x7d) {
            r.take();
            return el;
        }
        if (c === 0x7b) {
            el.children.push(decodeElement(r, depth + 1));
            continue;
        }
        if (c === BSLASH && (r.buf[r.i + 1] === 0x62 || r.buf[r.i + 1] === 0x42)) {
            // §5.1: a letter after the escape byte introduces a command. The
            // reserved bytes are all punctuation, so the two cannot collide.
            el.children.push(readBinary(r));
            continue;
        }
        const text = readEscaped(r, [0x7b, 0x7d]);
        if (text.length) el.children.push(text);
    }
}

function decode(buf) {
    const r = new Reader(Buffer.from(buf));
    const el = decodeElement(r, 0);
    if (!r.eof()) throw new XjarchiveError(`trailing bytes at offset ${r.i}`);
    return el;
}

function decodeDocument(buf) {
    const r = new Reader(Buffer.from(buf));
    const out = [];
    while (!r.eof()) out.push(decodeElement(r, 0));
    return out;
}

module.exports = {
    Element,
    XjarchiveError,
    IMPLICIT_TAG,
    MAX_DEPTH,
    encode,
    decode,
    encodeDocument,
    decodeDocument,
    chooseForm,
};
